File: RoadRunnerJS/src/jsMain/kotlin/trajectory/constraints/MinVelocityConstraint.kt
package trajectory.constraints

import geometry.Pose2d
import kotlin.js.JsExport

/**
 * Composite constraint representing the minimum of its constituent velocity constraints.
 */
@JsExport
class MinVelocityConstraint(
    private val constraints: Array<TrajectoryVelocityConstraint>
) : TrajectoryVelocityConstraint() {
    override fun get(s: Double, pose: Pose2d, deriv: Pose2d, baseRobotVel: Pose2d) =
        constraints.minOf { it[s, pose, deriv, baseRobotVel] }
}

File: src/roadrunnerjs/trajectory/constraints/MinVelocityConstraint.ts
import { Pose2d } from "roadrunnerjs/geometry";
import { TrajectoryVelocityConstraint } from "./TrajectoryVelocityConstraint";
/**
 * Composite constraint representing the minimum of its constituent velocity constraints.
 */
export class MinVelocityConstraint implements TrajectoryVelocityConstraint {
  constructor(private readonly constraints: TrajectoryVelocityConstraint[]) {}

  get(s: number, pose: Pose2d, deriv: Pose2d, baseRobotVel: Pose2d) {
    return Math.min(
      ...this.constraints.map((e) => e.get(s, pose, deriv, baseRobotVel))
    );
  }
}
